consensus.go

```diff
@@ -0,0 +1,235 @@
+	wallet := mw.WalletWithID(walletID)
+	if wallet == nil {
+		return nil, errors.New(ErrNotExist)
+	}
+	wal := wallet.Internal()
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return nil, fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	version, agendas := getAllAgendas(wal.ChainParams())
+	resp := &GetVoteChoicesResult{
+		Version: version,
+		Choices: make([]AgendaVoteChoice, len(agendas)),
+	}
+
+	ctx := wallet.shutdownContext()
+	choices, _, err := wal.AgendaChoices(ctx, ticketHash)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := range choices {
+		agenda := mw.agenda(agendas, choices[i].AgendaID)
+		resp.Choices[i] = AgendaVoteChoice{
+			AgendaID:          choices[i].AgendaID,
+			AgendaDescription: agenda.Vote.Description,
+			ChoiceID:          choices[i].ChoiceID,
+			ChoiceDescription: "", // Set below
+		}
+		for _, choice := range agenda.Vote.Choices {
+			if choices[i].ChoiceID == choice.Id {
+				resp.Choices[i].ChoiceDescription = choice.Description
+				break
+			}
+		}
+	}
+
+	return resp, nil
+}
+
+func (mw *MultiWallet) agenda(agendas []chaincfg.ConsensusDeployment, agendaID string) *chaincfg.ConsensusDeployment {
+	for _, agenda := range agendas {
+		if agenda.Vote.Id == agendaID {
+			return &agenda
+		}
+	}
+
+	return nil
+}
+
+// SetVoteChoice sets a voting choice for the specified agenda using the
+// specified ticket hash controlled by a VSP.
+//
+// If a ticket hash isn't provided, the vote choice is set for all tickets
+// controlled by the set VSP.
+func (mw *MultiWallet) SetVoteChoice(walletID int, vspHost string, vspPubKey []byte, agendaID, choiceID, hash string, passphrase []byte) error {
+	wallet := mw.WalletWithID(walletID)
+	if wallet == nil {
+		return errors.New(ErrNotExist)
+	}
+	wal := wallet.Internal()
+
+	if vspHost == "" {
+		return fmt.Errorf("request requires a vspHost but no vspHost was provided")
+	}
+
+	// Test to ensure the passphrase is correct, and the voting is authorized.
+	err := wallet.UnlockWallet(passphrase)
+	if err != nil {
+		return translateError(err)
+	}
+	defer wallet.LockWallet()
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	choice := w.AgendaChoice{
+		AgendaID: agendaID,
+		ChoiceID: choiceID,
+	}
+
+	ctx := wallet.shutdownContext()
+	_, err = wal.SetAgendaChoices(ctx, ticketHash, choice)
+	if err != nil {
+		return err
+	}
+
+	vspClient, err := wallet.VSPClient(vspHost, vspPubKey)
+	if err != nil {
+		return err
+	}
+
+	// If ticket hash is provided, then set vote choice for the selected ticket.
+	if ticketHash != nil {
+		vspTicketInfo, err := wal.VSPTicketInfo(ctx, ticketHash)
+		if err != nil {
+			return err
+		}
+		// Register the provided ticket hash with the VSP client.
```

Comment:
you are not registering a ticket hash here... you are starting/initializing the vsp client. The registration is done on line 132.

Reply:
from below that comment, the steps to register a ticket hash is started, the main vsp client is registered on line 116.

To set the vote choices for a particular ticket hash, the vsp client needs to be reinitialized, using the vsp host and pub key for the selected ticket, that's why the comment starts from that line

Reply:
What is being done here is setting the vote choice for the specified ticket with the vsp, NOT registering a ticket hash with the vsp client. Registering a ticket hash with a vsp client is done after the ticket is purchased and the fee is paid.